File: projects/Bouncy Ball/scripts/Paddle.lua
BlankE.addEntity("Paddle")

function Paddle:init()
	-- add the image of a paddle
	self.img_paddle = Image("paddle")
	-- change the image offset to the center of the image
	self.img_paddle.xoffset = self.img_paddle.width / 2
	self.img_paddle.yoffset = self.img_paddle.height / 2
	-- give the Paddle friction so it won't move in one direction forever
	self.friction = 0.4
	-- add a hitbox
	self:addShape("main","rectangle",{0,0,self.img_paddle.width,self.img_paddle.height})
end

function Paddle:update(dt)
	-- determine how fast it will move
	local move_spd = 500
	-- watch for key presses and move the paddle in the direction the player presses
	if Input("move_left").pressed then
		self.hspeed =	-move_spd
	end
	if Input("move_right").pressed then 
		self.hspeed =	move_spd 
	end 
	if Input("move_down").pressed then
		self.vspeed =	move_spd 
	end
	if Input("move_up").pressed then
		self.vspeed =	-move_spd
	end
	-- check if it is out of bounds horizontally
	-- 'teleport' it to the other side of the screen, if it goes out of bounds
	if self.x > game_width then
		self.x = 0
	end
	if self.x < 0 then
		self.x = game_width
	end
	-- move the paddle image
	self.img_paddle.x = self.x 
	self.img_paddle.y = self.y
end

function Paddle:explode()
	-- set a flag and check it so that the paddle can't explode again
	if self.exploded then return end
	self.exploded = true
	-- break the paddle image into pieces 
	self.img_paddle_pcs = self.img_paddle:chop(8,5)
	-- throw them in random directions
	local opp_direction = self.img_paddle.angle + 180
	self.img_paddle_pcs:forEach(function(piece)
		local direction = randRange(0, 360)
		piece.hspeed = direction_x(direction, 20)
		piece.vspeed = direction_y(direction, 20)
	end)
	Signal.emit("paddle_explode")
	-- remove the hitbox so that it seems like it's actually exploded and gone
	self:removeShape("main")
	-- after some time, destroy the object
	Timer.after(3, function()
		self:destroy()
	end)
end

function Paddle:draw()
	-- draw the pieces
	-- hide the old image if it's exploded
	if self.img_paddle_pcs then
		self.img_paddle_pcs:call('draw')
	else
		self.img_paddle:draw()
	end
end

File: projects/tut1-bouncyball/scripts/Paddle.js
class Paddle extends Entity {
    init () {
		this.addSprite("paddle")
		this.sprite_align = "center"
		this.sprite_scale.set(0.5);
		this.friction = 0.4;
		this.addShape("main","rect")
		// explode on contact with a missile
		this.onCollision['main'] = (other) => {
			if (other.tag == "Missile") {
				other.parent.destroy()
				this.explode()
			}
			if (other.tag == "Ball")
				console.log("oh wow")
		}
    }
    update (dt) {
		// Move the player presses keys
		let move_spd = 5;
		if (Input("move_left").pressed)
			this.hspeed = -move_spd;
		if (Input("move_right").pressed)
			this.hspeed = move_spd;
		if (Input("move_up").pressed)
			this.vspeed = -move_spd;
		if (Input("move_down").pressed)
			this.vspeed = move_spd;
		// 'teleport' it to the other side of the screen, if it goes out of bounds
		if (this.x > Game.width)
			this.x = 0;
		if (this.x < 0)
			this.x = Game.width;
		if (this.y < 0)
			this.y = Game.height;
		if (this.y > Game.height)
			this.y = 0;
    }
	explode () {
		if (this.exploded) return;
		this.exploded = true;
		
		Event.emit('paddle_explode')
		
		let paddle_bits = this.sprites['paddle'].chop(5,8)
		paddle_bits.forEach(b => {
			let direction = Util.rand_range(0,360)
			b.hspeed = Util.direction_x(direction, 10)
			b.vspeed = Util.direction_y(direction, 10)
		})
		this.destroy();
	}
}

TestScene({
	onStart (s) {
		Input.set('explosion','space')
		new Ball();
		s.pad = new Paddle();
		s.pad.y = Game.height/2;
		s.pad.x = s.pad.sprite_width/2;
		let ball = new Ball();
		ball.x = 40
	}, 
	onUpdate (s, dt) {
		if (Input('explosion').released) {
			s.pad.explode()
		}
	}
})